k2/csrc/aux_labels.cc

```diff
@@ -6,14 +6,89 @@
 
 #include "k2/csrc/aux_labels.h"
 
+#include <algorithm>
 #include <numeric>
+#include <utility>
 #include <vector>
 
 #include "glog/logging.h"
 #include "k2/csrc/fsa.h"
+#include "k2/csrc/fsa_util.h"
+#include "k2/csrc/properties.h"
+
+namespace {
+
+/*
+  This function counts how many extra states we need to create for each state in
+  the input FSA when we invert an FSA. Generally, if an entering arc of state
+  `i` in the input FSA has n olabels, then we need to create n-1 extra states
+  for state `i`.
+
+  @param [in] fsa_in      Input FSA
+  @param [in] labels_in   Aux-label sequences for the input FSA
+  @param [out] expanded_states  For state `i` in `fsa_in`, we need to create
+                                extra `expaned_states[i]` states in the output
+                                inverted FSA.
+*/
+static void CountExpandedStatesNum(const k2::Fsa &fsa_in,
+                                   const k2::AuxLabels &labels_in,
+                                   std::vector<int32_t> *expanded_states) {
+  CHECK_EQ(expanded_states->size(), fsa_in.NumStates());
+  auto &states = *expanded_states;
+  for (int32_t i = 0; i != fsa_in.arcs.size(); ++i) {
+    const auto &arc = fsa_in.arcs[i];
+    int32_t pos_start = labels_in.start_pos[i];
+    int32_t pos_end = labels_in.start_pos[i + 1];
+    states[arc.dest_state] += std::max(0, pos_end - pos_start - 1);
+  }
+}
+
+/*
+  Map the state in the input FSA to state in the output inverted FSA.
+
+  @param [in] expanded_states  Output of function `CountExpandedStatesNum`
+                               which gives how many extra states we need
+                               to create for each state in the input FSA.
+  @param [out] state_map       Map state `i` in the input FSA to state
+                               `state_map[i]` in the output FSA.
+  @param [in] state_ids        state_id[i] == state_map[i-1], it will be
```

Comment:
This documentation is not clear.  If `state_id[i]` were equal `state_map[i-1]`, why is there a need for this?

Reply:
`state_map` will be not changed after the function call, while the `state_id[i]` will be increaed by one each time when we number extra states created for state `i`, to get different state index for those extra states.
 Will add comments.

Reply:
If it will be changed then you should mark it [in,out].  Also, when you say `state_id[i] == state_map[i-1],`, is that true at entry or at exit?

Reply:
.. if you talk about this being changed *after* this function call (i.e. outside the function, by calling code), then that probably doesn't belong in this comment.  If you mean being changed *by* this function call, you need to be very clear about the values at entry and exit.

Reply:
I will set `state_id[0]` to 0, to it will just be `[out]`.